archive/100-theorems-list/45_partition.lean

```diff
@@ -0,0 +1,681 @@
+    subst this,
+    apply t,
+    refl }
+end
+
+def indicator_series (α : Type*) [semiring α] (f : set ℕ) : power_series α :=
+power_series.mk (λ n, if f n then 1 else 0)
+
+lemma coeff_indicator (f : set ℕ) [semiring α] (n : ℕ) :
+  coeff α n (indicator_series _ f) = if f n then 1 else 0 :=
+coeff_mk _ _
+lemma coeff_indicator_pos (f : set ℕ) [semiring α] (n : ℕ) (h : f n):
+  coeff α n (indicator_series _ f) = 1 :=
+by rw [coeff_indicator, if_pos h]
+lemma coeff_indicator_neg (f : set ℕ) [semiring α] (n : ℕ) (h : ¬f n):
+  coeff α n (indicator_series _ f) = 0 :=
+by rw [coeff_indicator, if_neg h]
+lemma constant_coeff_indicator (f : set ℕ) [semiring α] :
+  constant_coeff α (indicator_series _ f) = if f 0 then 1 else 0 :=
+by rw [← coeff_zero_eq_constant_coeff_apply, coeff_indicator]
+
+lemma two_series (i : ℕ) [semiring α] :
+  (1 + (X : power_series α)^i.succ) = indicator_series α {0, i.succ} :=
+begin
+  ext,
+  simp only [coeff_indicator, coeff_one, add_monoid_hom.map_add, coeff_X_pow,
+             ← @set.mem_def _ _ {0, i.succ}, set.mem_insert_iff, set.mem_singleton_iff],
+  cases n,
+    simp [(nat.succ_ne_zero i).symm],
+  simp [nat.succ_ne_zero n],
+end
+
+lemma num_series' [field α] (i : ℕ) :
+  (1 - (X : power_series α)^(i+1))⁻¹ = indicator_series α (λ k, i + 1 ∣ k) :=
+begin
+  rw power_series.inv_eq_iff,
+  { ext,
+    cases n,
+    { simp [mul_sub, zero_pow, constant_coeff_indicator] },
+    { rw [coeff_one, if_neg (nat.succ_ne_zero n), mul_sub, mul_one, add_monoid_hom.map_sub,
+          coeff_indicator],
+      simp_rw [coeff_mul, coeff_X_pow, coeff_indicator, boole_mul, sum_ite, filter_filter,
+               sum_const_zero, add_zero, sum_const, nsmul_eq_mul, mul_one, sub_eq_iff_eq_add,
+               zero_add, filter_congr_decidable],
+      symmetry,
+      split_ifs,
+      { suffices : ((nat.antidiagonal n.succ).filter (λ (a : ℕ × ℕ), i + 1 ∣ a.fst ∧ a.snd = i + 1)).card = 1,
+          rw this, norm_cast,
+        rw card_eq_one,
+        cases h with p hp,
+        refine ⟨((i+1) * (p-1), i+1), _⟩,
+        ext ⟨a₁, a₂⟩,
+        simp only [mem_filter, prod.mk.inj_iff, nat.mem_antidiagonal, mem_singleton],
+        split,
+        { rintro ⟨_, ⟨a, rfl⟩, rfl⟩,
+          refine ⟨_, rfl⟩,
+          rw [nat.mul_sub_left_distrib, ← hp, ← a_left, mul_one, nat.add_sub_cancel] },
+        { rintro ⟨rfl, rfl⟩,
+          cases p,
+            rw mul_zero at hp, cases hp,
+          rw hp,
+          simp [nat.succ_eq_add_one, mul_add] } },
+      { suffices : (filter (λ (a : ℕ × ℕ), i + 1 ∣ a.fst ∧ a.snd = i + 1) (nat.antidiagonal n.succ)).card = 0,
+          rw this, norm_cast,
+        rw card_eq_zero,
+        apply eq_empty_of_forall_not_mem,
+        simp only [prod.forall, mem_filter, not_and, nat.mem_antidiagonal],
+        rintro _ h₁ h₂ ⟨a, rfl⟩ rfl,
+        apply h,
+        simp [← h₂] } } },
+  { simp [zero_pow] },
+end
+
+lemma card_eq_of_bijection {β : Type*} {s : finset α} {t : finset β}
```

Comment:
Is this functionally the same as `finset.card_congr`?

Reply:
You're right - I *knew* that existed somewhere but I just couldn't find it so I made my own! I'll amend to use that :)